bin/monitor

```diff
@@ -2,62 +2,141 @@
+  ThreadPrinter.run
+  Kernel.exit!
+end
 
-        loop do
-          sessions[r.id][:ssh_session].process
-        rescue => ex
-          Clog.emit("Processing SSH session is failed. Trying to reestablish the connection") { {health_monitor_ssh_failure: {ubid: r.ubid, exception: Util.exception_to_hash(ex)}} }
-          break
+consumer = Thread.new do
+  loop do
+    Clog.emit("Health checks couldn't be completed.") { {health_monitor_overload: {incomplete_health_check_count: (Thread.list.count - 3) / 2}} } if Thread.list.count > 3
+    mutex.synchronize do
+      keys_to_delete = []
+      resources.each do |r_id, r|
+        while Thread.list.count + 2 > Config.max_monitor_threads
+          sleep 1
         end
-      rescue => ex
-        Clog.emit("Establishing the SSH session is failed") { {health_monitor_reestablish_ssh_failure: {ubid: r.ubid, exception: Util.exception_to_hash(ex)}} }
-
-        begin
-          r.reload
-        rescue Sequel::NoExistingObject
-          Clog.emit("Resource is deleted") { {health_monitor_resource_deleted: {ubid: r.ubid}} }
-          sessions.delete(r.id)
-          ssh_threads.delete(r.id)
-          pulse_threads.delete(r.id)&.kill
-          break
+
+        # If the SSH thread didn't stopped yet, it means that it didn't get signal
+        # from the pulse thread to exit yet. That means the pulse thread is also
+        # still running. We should skip this resource, its health check from previous
+        # iteration is not completed yet.
+        # However, there is one edge case we need to handle. If the pulse thread is
+        # blocked, it will never send the signal to the ssh thread to exit. In that
+        # case, we will not be able to monitor that resource anymore. In such case,
+        # we will shutdown the ssh session, which would hopefully wake up the pulse
+        # thread (with an exception) and it will exit.
+        if r[:ssh_thread_status] != ThreadStatus::STOPPED
+          if r[:pulse_thread_started_at] + PULSE_THREAD_TIMEOUT < Time.now
+            r[:session][:ssh_session].shutdown!
+          end
+
+          next
         end
 
-        sleep 5
-      end
-    end
+        # We have limits on the number of threads we can have simultaneously. Thus,
+        # both ssh and pulse threads exits after completing the health check of a
+        # given resource. This opens up space for another resource to create its own
+        # threads. Alternative would be having a thread pool, but that would require
+        # non-trivial amount coordination to start pulse and ssh threads together.
+        # Also creating a new thread is not very expensive; creating 1000 threads
+        # takes about 0.025 seconds in my (random developer) laptop, so overhead is
+        # negligible.
+        Thread.new do
+          until r[:ssh_thread_status] == ThreadStatus::STOPPING
+            begin
+              r[:session] ||= r[:model].reload.init_health_monitor_session
```

Comment:
I still don't quite understand precisely what this reload is doing, is it basically to help the thing terminate once deleted, as well as picking up a snapshot of fields to make a session? 

But if there's an existing session...then there is no check of the record anyway...

Reply:
So, this reload only runs if the r[:session] is nil, which only happens if previous pulse check is failed and we set it to nil for re-establishing the session. One reason for pulse check failure is the resource deletion, and reload catches that and terminates the thread and also removes the resource from the monitor list.